hir/src/asm/builder.rs

```diff
@@ -404,24 +404,99 @@ impl<'a> MasmOpBuilder<'a> {
+        let cond = self.stack.pop().expect("operand stack is empty");
+        assert_eq!(
+            cond,
+            Type::I1,
+            "expected while.true condition to be a boolean value"
+        );
+        let out_stack = self.stack.clone();
+        IfTrueBuilder {
+            dfg: self.dfg,
+            asm: self.asm,
+            in_stack: self.stack,
+            out_stack,
+            ip: self.ip,
+            then_blk: None,
+            else_blk: None,
+        }
     }
 
-    /// Pops a boolean value from the stack, and executes the given block if it is true,
-    /// otherwise it is skipped. The given block will continue to execute for as long as
-    /// the top value on the stack at the end of the block is true.
-    pub fn while_true(self, body: MasmBlockId) {
-        self.stack.drop();
-        self.asm.push(self.ip, MasmOp::While(body));
+    /// Begins construction of a `while.true` loop.
+    ///
+    /// A `while.true` pops a boolean value off the stack to use as the condition for
+    /// entering the loop, and will then execute the loop body for as long as the value
+    /// on top of the stack is a boolean and true. If the condition is not a boolean,
+    /// execution traps.
+    ///
+    /// NOTE: This function will panic if the top of the operand stack is not of boolean type
+    /// when called.
+    ///
+    /// Before finalizing construction of the loop body, you must ensure two things:
+    ///
+    /// 1. There is a value of boolean type on top of the operand stack
+    /// 2. The abstract state of the operand stack, assuming the boolean just mentioned
+    /// has been popped, must be consistent with the state of the operand stack when the
+    /// loop was entered, as well as if the loop was skipped due to the conditional being
+    /// false. The abstract state referred to here is the number, and type, of the elements
+    /// on the operand stack.
+    ///
+    /// Both of these are validated by [LoopBuilder], and a panic is raised if validation fails.
+    pub fn while_true(self) -> LoopBuilder<'a> {
+        let cond = self.stack.pop().expect("operand stack is empty");
+        assert_eq!(
+            cond,
+            Type::I1,
+            "expected while.true condition to be a boolean value"
+        );
+        let out_stack = self.stack.clone();
+        let body = self.asm.create_block();
+        LoopBuilder {
+            dfg: self.dfg,
+            asm: self.asm,
+            in_stack: self.stack,
+            out_stack,
+            ip: self.ip,
+            body,
+            style: LoopType::While,
+        }
     }
 
-    /// Repeatedly executes `body`, `n` times.
-    pub fn repeat(self, n: u8, body: MasmBlockId) {
-        self.asm.push(self.ip, MasmOp::Repeat(n, body));
+    /// Begins construction of a `repeat` loop, with an iteration count of `n`.
+    ///
+    /// A `repeat` instruction requires no operands on the stack, and will execute the loop body `n` times.
+    ///
```

Comment:
Suggested addition: "The abstract state of the operand stack must be consistent when the loop body is entered, as well as when the loop is completed."